crates/iceberg/src/arrow/reader.rs

```diff
@@ -186,4 +216,399 @@ impl ArrowReader {
         }
     }
+
+    fn get_row_filter(&self, parquet_schema: &SchemaDescriptor) -> Result<Option<RowFilter>> {
+        if let Some(predicates) = &self.predicates {
+            let field_id_map = self.build_field_id_map(parquet_schema)?;
+
+            // Collect Parquet column indices from field ids
+            let mut collector = CollectFieldIdVisitor { field_ids: vec![] };
+            visit_predicate(&mut collector, predicates).unwrap();
+            let column_indices = collector
+                .field_ids
+                .iter()
+                .map(|field_id| {
+                    field_id_map.get(field_id).cloned().ok_or_else(|| {
+                        Error::new(ErrorKind::DataInvalid, "Field id not found in schema")
+                    })
+                })
+                .collect::<Result<Vec<_>>>()?;
+
+            // Convert BoundPredicates to ArrowPredicates
+            let mut converter = PredicateConverter {
+                columns: &column_indices,
+                projection_mask: ProjectionMask::leaves(parquet_schema, column_indices.clone()),
+                parquet_schema,
+                column_map: &field_id_map,
+            };
+            let arrow_predicate = visit_predicate(&mut converter, predicates)?;
+            Ok(Some(RowFilter::new(vec![arrow_predicate])))
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Build the map of field id to Parquet column index in the schema.
+    fn build_field_id_map(&self, parquet_schema: &SchemaDescriptor) -> Result<HashMap<i32, usize>> {
+        let mut column_map = HashMap::new();
+        for (idx, field) in parquet_schema.columns().iter().enumerate() {
+            let field_type = field.self_type();
+            match field_type {
+                ParquetType::PrimitiveType { basic_info, .. } => {
+                    if !basic_info.has_id() {
+                        return Err(Error::new(
+                            ErrorKind::DataInvalid,
+                            format!(
+                                "Leave column {:?} in schema doesn't have field id",
+                                field_type
+                            ),
+                        ));
+                    }
+                    column_map.insert(basic_info.id(), idx);
+                }
+                ParquetType::GroupType { .. } => {
+                    return Err(Error::new(
+                        ErrorKind::DataInvalid,
+                        format!(
+                            "Leave column in schema should be primitive type but got {:?}",
+                            field_type
+                        ),
+                    ));
+                }
+            };
+        }
+
+        Ok(column_map)
+    }
+}
+
+/// A visitor to collect field ids from bound predicates.
+struct CollectFieldIdVisitor {
+    field_ids: Vec<i32>,
+}
+
+impl BoundPredicateVisitor for CollectFieldIdVisitor {
```

Comment:
#334 has got merged, do you mind to rewrite this?

Reply:
Okay, let me take a look, not sure if it is completely suitable for the usage here, especially for `PredicateConverter`. But let me try it.

Reply:
Rewrote it using the API.